app/services/courses/validate_unique_course_on_provider_service.rb

```diff
@@ -0,0 +1,21 @@
+module Courses
+  class ValidateUniqueCourseOnProviderService
```

Comment:
Should this line, given this class is called `ValidateUniqueXXX`

`@course.errors.add(:base, :duplicate)` sit in here so controller is not subjected to change?

Reply:
Interesting point. I'd also ask why not use a custom validation within the `Course` model?

Reply:
I mainly wanted to have it separate as I think we should be more explicit on how we determine whether or not a course is unique. 

This could potentially be moved to a `validates_with` method on the course though now that I think about it (this would pull in Sau's suggestion also)

Reply:
Ya, that's what I meant, sorry I wasn't clear. Basically create a custom validator and use `validates_with`, using the existing service-like mechanism.

Reply:
:+1: sounds good